File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int index = binarySearch(nums, target);
        if(index == -1){
            return new int[] {-1, -1};
        }
        int left = index, right = index;
        while(left > 0 && nums[left - 1] == nums[index]){
            left --;
        }
        while(right < nums.length - 1 && nums[right + 1] == nums[index]){
            right ++;
        }
        return new int[] {left, right};
    }
    
    public int binarySearch(int[] nums, int target){
        if (nums.length < 1 || target < nums[0] || target > nums[nums.length-1]){
            return -1;
        }
        int left = 0, right = nums.length-1;
        while(left <= right){
            int mid = left + ((right - left) >>1);
            if(nums[mid] == target){
                return mid;
            }else if(nums[mid] > target){
                right = mid-1;
            }else{
                left = mid +1;
            }
        }
        return -1;
    }
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        def findLeftBound(nums, target):
            left, right = 0, len(nums) - 1
            leftBound = -2
            while left <= right:
                middle = left + (right - left) // 2
                if nums[middle] >= target:
                    right = middle - 1
                    leftBound = right
                else:
                    left = middle + 1
            print(leftBound)
            return leftBound
        
        def findRightBound(nums, target):
            left, right = 0, len(nums) - 1
            rightBound = -2
            while left <= right:
                middle = left + (right - left) // 2
                if nums[middle] <= target:
                    left = middle + 1
                    rightBound = left
                else:
                    right = middle -1
            print(rightBound)
            return rightBound
        
        left = findLeftBound(nums, target)
        right = findRightBound(nums, target)
        
        if left == -2 or right == -2:
            return [-1, -1]
        elif right - left > 1:
            return [left + 1, right - 1]
        else:
            return [-1, -1]
        